### PR TITLE
PointConv + Linear Layers Fusion and using KNN Inverse Indices

### PR DESCRIPTION
**Pull request by Logeswaran123**
This branch changes includes:

1. PointConv + Linear Layer fused kernels for forward and backward.
2. KNN Inverse Indices computation kernel.
3. Optimized fused backward kernel using Inverse Indices.
4. Optimized fused cutlass forward kernel.
5. Tests for Performance analysis.
6. Integration with existing model architecture.
7. To enable Optimized kernels, set `USE_CUDA_KERNEL: True` and `PCONV_OPT: True` in the config file. Refer configs/configPCF_Opt_10cm.yaml

Cutlass forward kernel is not yet integrated into model architecture (more optimization work to be done here).

Below Profiles are with H100 GPU, CUDA 11.8, Cutlass 3.x.

For kernel profilings, refer this log. This is for single run, so the kernel launch overhead is significant. But, it's good to look at grid and block sizes for all the kernels in a single place.
[pconv_nvtx_log.txt](https://github.com/user-attachments/files/19699228/pconv_nvtx_log.txt)

For plots profiling runtime and memory usage averaged across 50 runs:
Profile of Unfused vs. Fused vs. Fused using KNN Inverse Indices kernels:
pconv_benchmark_results
pconv_benchmark_speedup

Profile of KNN Inverse Indices kernel:
knn_inv_benchmark_results
knn_inv_k_scaling

To-Dos:
1. Optimize memory usage of cutlass forward kernel.
2. Adapt Fused backward Pure CUDA kernel to cutlass.

### CODE REVIEW

File: cpp_wrappers/cpp_pcf_kernel/pcf_cuda.cpp [hunk 1]
```diff
@@ -6,6 +6,19 @@
...
+    torch::Tensor linear_weights,
+    torch::Tensor linear_bias);
+
+
+std::vector<torch::Tensor> knn_inverse_cuda_forward(
+    torch::Tensor neighbor_inds,
+    const int total_points);
+
 // TODO: For now, not attempting to fuse the next downsampling linear layer because hand-written matmul 
 //       cannot compare with optimized gemm kernels. May explore using CUTLASS for that at a later time.
```
Comment (pvskand) [inline cpp_wrappers/cpp_pcf_kernel/pcf_cuda.cpp]: Can remove this comment since this is indeed the CUTLASS implementation.

File: README.md [hunk 2]
```diff
@@ -62,6 +62,8 @@ setup.sh
 
 7. To enable GPU-based KNN computation, set `post_knn: true` in the configuration file. Refer to [`train_ScanNet_DDP_WarmUP.py`](./train_ScanNet_DDP_WarmP.py) for an example implementation, Load data (`features`, `pointclouds`, `target`, `norms`, `points_stored`), Compute kNN edges using `compute_knn_packed()`, Prepare edges for further processing using `prepare()`
 
+8. To enable optimized PointConv CUDA kernels for saving peak GPU memory usage, set `USE_CUDA_KERNEL: True` and `PCONV_OPT: True` in the configuration file. Refer to ['configPCF_Opt_10cm.yaml'](./configs/configPCF_Opt_10cm.yaml). Use the fused PointConv and Linear layer forward and backward kernels, Compute kNN Inverse mapping indices for all the edges, use the inverse indices in the backward kernel to compute gradients.
```
Comment (pvskand) [inline README.md:65]: `PCONV_OPT` is with CUTLASS? If so mention that.

File: configs/configPCF_Opt_10cm.yaml [hunk 3]
```diff
@@ -0,0 +1,80 @@
+# See get_default_configs() in train_ScanNet_DDP_Warmup.py
+# and get_default_configs() in model_architecture.py for the meanings of those configs
+USE_CUDA_KERNEL: True
+PCONV_OPT:      True
```
Comment (pvskand) [inline configs/configPCF_Opt_10cm.yaml:4]: Also wondering if we can replace this variable name with `USE_CUTLASS` to be consistent with `USE_CUDA_KERNEL`?

Any potential problems that you foresee with this @Logeswaran123?

File: cpp_wrappers/cpp_pcf_kernel/profile.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+
+        torch.cuda.nvtx.range_push(f"{impl['name']}_BACKWARD")
+        impl["backward"](out)
+        torch.cuda.synchronize()
+        torch.cuda.nvtx.range_pop()
+
+        after_backward = memory_usage_stats()
+        peak_after_backward = torch.cuda.max_memory_allocated() / (1024 * 1024)
+
+        print_memory_usage("Backward pass", before_backward, after_backward)
```
Comment (pvskand) [inline cpp_wrappers/cpp_pcf_kernel/profile.py:220]: Can you also print out the following (and have an `assert` statement):

1. Error in `forward()` pass of `PConvLinear()`, `PConvLinearOpt()` and `PConvLinearCutlass` wrt to `PConv()`.
2. Similarly, error in the `backward()` pass of the three wrt to `PConv() backward()`.

File: cpp_wrappers/cpp_pcf_kernel/setup.py [hunk 5]
```diff
@@ -15,7 +15,12 @@
         CUDAExtension('pcf_cuda', [
             'pcf_cuda.cpp',
             'pcf_cuda_kernel.cu',
-        ], extra_compile_args={'nvcc': ['-L/usr/local/cuda/lib64 -lcudadevrt -lcudart']})
+        ],
+        include_dirs=[
+            '/nfs/stak/users/sivakuml/hpc-memory/cutlass/cutlass/include',
+            '/nfs/stak/users/sivakuml/hpc-memory/cutlass/cutlass/tools/util/include',
```
Comment (pvskand) [inline cpp_wrappers/cpp_pcf_kernel/setup.py:21]: Hard coded directories -- need to automate them.

File: util/common_util.py [hunk 6]
```diff
@@ -245,3 +245,55 @@ def replace_batchnorm(net):
             setattr(net, child_name, torch.nn.Identity())
         else:
             replace_batchnorm(child)
+
+
+def compute_knn_inverse(pointclouds, edges_self, edges_forward, edges_propagate):
```
Comment (pvskand) [inline util/common_util.py:250]: Comment what this function does in detail.

File: train_ScanNet_DDP_WarmUP.py [hunk 7]
```diff
@@ -503,6 +511,11 @@ def train(train_loader, model, criterion, optimizer, epoch, scheduler):
 #                                'mAcc_train_batch': np.mean(intersection / (target + 1e-10)),
 #                                'allAcc_train_batch': accuracy,
 #                                'lr_train_batch': lr}
+
+        del inv_self, inv_forward, inv_propagate
+        del pred, loss, output, intersection, union, target
+        torch.cuda.empty_cache()
```
Comment (pvskand) [inline train_ScanNet_DDP_WarmUP.py:517]: Are these lines necessary to ensure that memory doesn't blow up during training? If so, we need a work around this as we can't expect users to keep writing this every time they perform a forward+backward pass.

File: layers.py [hunk 8]
```diff
@@ -311,7 +311,10 @@ def forward(
             dense_xyz_norm,
             sparse_xyz=None,
             sparse_xyz_norm=None,
-            vi_features=None):
+            vi_features=None,
+            inv_neighbors=None,
+            inv_k=None,
+            inv_idx=None):
```
Comment (pvskand) [inline layers.py:317]: Comment of the function should include these 4 variables along with their sizes as well.

File: layers.py [hunk 9]
```diff
@@ -779,7 +817,10 @@ def forward(
             nei_inds,
             dense_xyz_norm=None,
             sparse_xyz=None,
-            sparse_xyz_norm=None):
+            sparse_xyz_norm=None,
+            inv_neighbors=None,
+            inv_k=None,
+            inv_idx=None):
```
Comment (pvskand) [inline layers.py:823]: Comment of the function should include these 4 variables along with their sizes as well.

File: layers.py [hunk 10]
```diff
@@ -935,7 +1005,11 @@ def forward(
             dense_xyz,
             dense_xyz_norm,
             dense_feats=None,
-            vi_features=None):
+            vi_features=None,
+            inv_neighbors=None,
+            inv_k=None,
+            inv_idx=None
+            ):
```
Comment (pvskand) [inline layers.py:1012]: Comment of the function should include these 4 variables along with their sizes as well.

File: layer_utils.py [hunk 11]
```diff
@@ -30,6 +30,52 @@ def index_points(points, idx):
     return new_points
 
 
+class PConvLinearOptFunction(torch.autograd.Function):
```
Comment (pvskand) [inline layer_utils.py:33]: Explaining what the cuda kernel does briefly in a couple of sentences would be good here. That way, people who don't want to dig into the kernel but want to know how this particular implementation is more optimized than the naive PointConv can go through it.

File: layer_utils.py [hunk 12]
```diff
@@ -30,6 +30,52 @@ def index_points(points, idx):
...
+        grad_output = grad_output.contiguous()
+
+        grads = pcf_cuda.pconv_linear_opt_backward(
+            grad_output, input_feat, inverse_neighbors, inverse_k, 
+            inverse_idx, neighbor_inds, weightnet, additional_features,
+            linear_weights, pconv_output)
+
+        return grads[0], None, None, None, None, grads[1], grads[2], grads[3], grads[4]
+
+class PConvLinearOpt(torch.nn.Module):
```
Comment (pvskand) [inline layer_utils.py:63]: Same comment as above: 
Explaining what the cuda kernel does briefly in a couple of sentences would be good here. That way, people who don't want to dig into the kernel but want to know how this particular implementation is more optimized than the naive PointConv can go through it.

**Review by pvskand — CHANGES_REQUESTED** [review]
Overall looks good. Fantastic work! Major comments are on the following:
1. Commenting the code
2. Tests for accuracy of the `forward` and `backward` when you are profiling. You profile the code at multiple places -- so unifying it and having all the profiling + accuracy check between the optimized PointConvs and original PointConv at one place would be good.

File: cpp_wrappers/cpp_pcf_kernel/profile.py (hunk 4 above)
Comment (Logeswaran123) [reply cpp_wrappers/cpp_pcf_kernel/profile.py:220]: This file is only to profile the runtime and memory usage stats for all the kernels using nsys profile.
`nsys profile --stats=true --force-overwrite=true --cuda-memory-usage=true --trace=cuda,nvtx,osrt --output=pconv_profile python3 cpp_wrappers/cpp_pcf_kernel/profile.py &> log.txt`

The error is forward and backward w.r.t. baseline versions (PConv() and PConv() backward()) are in test_kernels.py

File: configs/configPCF_Opt_10cm.yaml (hunk 3 above)
Comment (Logeswaran123) [reply configs/configPCF_Opt_10cm.yaml:4]: I prefer to use PCONV_OPT to indicate optimized versionsm because CUTLASS GEMMS are used only in forward and a custom CUDA kernel is used in backward.
